File: mobile/lib/data/models/rfq.dart
class RFQ {
  final String id;
  final String rfqNumber;
  final String title;
  final String? description;
  final String status;
  final String? deadline;
  final String? createdAt;
  final int? budgetCents;
  final List<RFQLineItem> lineItems;
  final List<RFQBid> bids;

  const RFQ({
    required this.id,
    required this.rfqNumber,
    required this.title,
    this.description,
    required this.status,
    this.deadline,
    this.createdAt,
    this.budgetCents,
    this.lineItems = const [],
    this.bids = const [],
  });

  factory RFQ.fromJson(Map<String, dynamic> json) {
    return RFQ(
      id: json['id'] ?? '',
      rfqNumber: json['rfq_number'] ?? json['rfqNumber'] ?? '',
      title: json['title'] ?? '',
      description: json['description'],
      status: json['status'] ?? '',
      deadline: json['deadline'],
      createdAt: json['created_at'],
      budgetCents: json['budget_cents'],
      lineItems:
          (json['line_items'] as List<dynamic>?)
              ?.map((e) => RFQLineItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      bids:
          (json['bids'] as List<dynamic>?)
              ?.map((e) => RFQBid.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class RFQLineItem {
  final String id;
  final String description;
  final double quantity;
  final String? unit;

  const RFQLineItem({
    required this.id,
    required this.description,
    required this.quantity,
    this.unit,
  });

  factory RFQLineItem.fromJson(Map<String, dynamic> json) {
    return RFQLineItem(
      id: json['id'] ?? '',
      description: json['description'] ?? '',
      quantity: (json['quantity'] ?? 0).toDouble(),
      unit: json['unit'],
    );
  }
}

class RFQBid {
  final String id;
  final String vendorId;
  final int totalCents;
  final int? deliveryDays;
  final String? notes;

  const RFQBid({
    required this.id,
    required this.vendorId,
    required this.totalCents,
    this.deliveryDays,
    this.notes,
  });

  factory RFQBid.fromJson(Map<String, dynamic> json) {
    return RFQBid(
      id: json['id'] ?? '',
      vendorId: json['vendor_id'] ?? '',
      totalCents: json['total_cents'] ?? 0,
      deliveryDays: json['delivery_days'],
      notes: json['notes'],
    );
  }
}

File: api/models/rfq.py
import uuid
from datetime import datetime
from typing import Optional
from decimal import Decimal

from sqlalchemy import (
    String,
    BigInteger,
    Integer,
    Numeric,
    DateTime,
    Text,
    ForeignKey,
    UniqueConstraint,
    CheckConstraint,
    Index,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from api.database import Base


class Rfq(Base):
    __tablename__ = "rfqs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False
    )
    rfq_number: Mapped[str] = mapped_column(
        String(50), unique=True, nullable=False
    )
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    pr_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("purchase_requests.id")
    )
    status: Mapped[str] = mapped_column(String(20), default="DRAFT")
    deadline: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    awarded_vendor_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=True
    )
    awarded_po_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("purchase_orders.id"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow
    )

    __table_args__ = (
        CheckConstraint(
            "status IN ('DRAFT','OPEN','CLOSED','AWARDED','CANCELLED')",
            name="chk_rfq_status",
        ),
    )


class RfqLineItem(Base):
    __tablename__ = "rfq_line_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    rfq_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("rfqs.id", ondelete="CASCADE"),
        nullable=False,
    )
    description: Mapped[str] = mapped_column(Text, nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    specifications: Mapped[Optional[str]] = mapped_column(Text)

    __table_args__ = (
        CheckConstraint("quantity > 0", name="chk_rfq_line_qty"),
    )


class RfqBid(Base):
    __tablename__ = "rfq_bids"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False
    )
    rfq_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("rfqs.id"), nullable=False
    )
    vendor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=False
    )
    total_cents: Mapped[int] = mapped_column(BigInteger, nullable=False)
    delivery_days: Mapped[Optional[int]] = mapped_column(Integer)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    score: Mapped[Optional[Decimal]] = mapped_column(Numeric(5, 2))
    submitted_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow
    )

    __table_args__ = (
        UniqueConstraint("rfq_id", "vendor_id", name="uq_rfq_bid_vendor"),
    )
